### PR TITLE
feature/COR-1435-hospitals-page-patient-influx-over-time-filter-null-values

### PR DESCRIPTION
**Pull request by VWSCoronaDashboard26**
## Summary

* added `countLeadingNullValues` utility and tests;
* added `trimLeadingNullValues` utility and tests;
* refactored `trimNullValues` to `trimTrailingNullValues`;
* implemented usage of new utilities on the 'Ziekenhuizen en de zorg' page;
* implemented small quality of life fix (persist state when toggling between the graphs on this page);

### Screenshots
The screenshot only show the  'Ziekenhuisinstroom' (or hospital scope) of the 'Instroom van covid-patiënten door de tijd heen' graph. However, it should be noted that this change is also applied for the 'Intensive care-instroom' (or ICU scope).
#### Before
<details>
<summary>Toggle before screenshots</summary>

COR-1435_before

</details>

#### After
<details>
<summary>Toggle after screenshots</summary>

COR-1435_after

</details>

### CODE REVIEW

File: packages/app/src/utils/count-leading-null-values.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import { findIndex } from 'lodash';
```
Comment (VWSCoronaDashboard28) [inline packages/app/src/utils/count-leading-null-values.ts:1]: Why not use `findIndex` which is available in JS? Not needed to import it from `lodash` I would say

File: packages/app/src/utils/count-leading-null-values.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+import { isFilled, isPresent } from 'ts-is-present';
+
+/**
+ * Return the number of leading null values from the given value array.
+ * If a property name is specified it is used to check each value item's
+ * property for null.
+ */
+export const countLeadingNullValues = <T>(values: T[], property?: keyof T) => {
+  const predicate = isPresent(property) ? (d: T) => isFilled(d[property]) : (d: T) => isFilled(d);
```
Comment (VWSCoronaDashboard28) [inline packages/app/src/utils/count-leading-null-values.ts:10]: Why is this called predicate? I'm not really sure what to understand by that variable name.

I am also not too big a fan of `(d: T)` what is `d` supposed to be?

File: packages/app/src/utils/count-trailing-null-values.ts [hunk 3]
```diff
@@ -6,10 +6,8 @@ import { isFilled, isPresent } from 'ts-is-present';
  * If a property name is specified it is used to check each value item's
  * property for null.
  */
-export function countTrailingNullValues<T>(values: T[], property?: keyof T) {
-  const predicate = isPresent(property)
-    ? (d: T) => isFilled(d[property])
-    : (d: T) => isFilled(d);
+export const countTrailingNullValues = <T>(values: T[], property?: keyof T) => {
```
Comment (VWSCoronaDashboard28) [inline packages/app/src/utils/count-trailing-null-values.ts:9]: Same comments here as made for `count-leading-bull-values`

File: packages/app/src/utils/trim-leading-null-values.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+ * checked for null on each item.
+ *
+ * @param values The given array
+ * @param property Optional property to check
+ * @returns The existing array without the leading null values
+ */
+export const trimLeadingNullValues = <T>(values: T[], property?: keyof T) => {
+  const numberOfLeadingNulls = countLeadingNullValues(values, property);
+
+  return values.filter((value) => !(values.indexOf(value) < numberOfLeadingNulls));
```
Comment (VWSCoronaDashboard28) [inline packages/app/src/utils/trim-leading-null-values.ts:15]: I'm a little confused here, so `countLeadingNullValues` returns the number of null values at the beginning of an array so for example if i returns 20, that means the first 20 items of the array contain null values? 

Wouldn't `values.slice(numberOfLeadingValues)` suffice?

File: packages/app/src/utils/trim-leading-null-values.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import { countLeadingNullValues } from './count-leading-null-values';
+
+/**
+ * Trims all the leading null values from the given array.
+ * When the property parameter is defined, this property will be
+ * checked for null on each item.
+ *
+ * @param values The given array
```
Comment (VWSCoronaDashboard28) [inline packages/app/src/utils/trim-leading-null-values.ts:8]: Could you perhaps change this @param values The given array to @param values - The given array? Same for the param below

File: packages/app/src/utils/trim-trailing-null-values.ts [hunk 6]
```diff
@@ -1,16 +1,16 @@
 import { countTrailingNullValues } from './count-trailing-null-values';
 
 /**
- * Right trims all of the null values from the given array.
- * When the property parameter is defined this property will be
+ * Trims all the trailing null values from the given array.
+ * When the property parameter is defined, this property will be
  * checked for null on each item.
  *
  * @param values The given array
```
Comment (VWSCoronaDashboard28) [inline packages/app/src/utils/trim-trailing-null-values.ts:8]: Could you perhaps change this @param values The given array to @param values - The given array? Same for the param below

File: packages/app/src/utils/count-leading-null-values.ts (hunk 2 above)
Comment (VWSCoronaDashboard26) [reply packages/app/src/utils/count-leading-null-values.ts]: It is called `predicate` because its value is used to affirm/assert if the passed in value meets what is expected.

I do agree on `d` being very cryptic, and seeing as it is an instance of `T`, one can conclude that a more logical name for it would be `value`. I have updated the scripts accordingly, so feel free to resolve this if you agree.

File: packages/app/src/utils/trim-leading-null-values.ts (hunk 4 above)
Comment (VWSCoronaDashboard26) [reply packages/app/src/utils/trim-leading-null-values.ts]: Exactly: `countLeadingNullValues` will find the first index of a filled value (a value that is asserted through the predicate). Anything before that does not meet the predicate and is therefore a `null` value.

I am 100% sure I had tried `values.slice(numberOfLeadingValues)` earlier, but that did not seem to work at the time. However, it works fine now so I have updated this script accordingly, so feel free to resolve this if you agree.

File: packages/app/src/utils/trim-leading-null-values.ts (hunk 5 above)
Comment (VWSCoronaDashboard26) [reply packages/app/src/utils/trim-leading-null-values.ts]: Good one! Applied this and added more docs to the other utils. Feel free to resolve this if you agree.

File: packages/app/src/utils/trim-trailing-null-values.ts (hunk 6 above)
Comment (VWSCoronaDashboard26) [reply packages/app/src/utils/trim-trailing-null-values.ts]: Good one! Applied this and added more docs to the other utils. Feel free to resolve this if you agree.

**Review by VWSCoronaDashboard28 — APPROVED** [review]

File: packages/app/src/utils/count-trailing-null-values.ts (hunk 3 above)
Comment (VWSCoronaDashboard26) [reply packages/app/src/utils/count-trailing-null-values.ts:13]: See other comments. Feel free to resolve this if you agree.

File: packages/app/src/utils/trim-leading-null-values.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+ * @param property Optional property to check
+ * @returns The existing array without the leading null values
+ */
+export const trimLeadingNullValues = <T>(values: T[], property?: keyof T) => {
+  const numberOfLeadingNulls = countLeadingNullValues(values, property);
```
Comment (VWSCoronaDashboard18) [inline packages/app/src/utils/trim-leading-null-values.ts:13]: Is there a reason you created two new functions? Instead of using the existing one with a double reverse for example?

**Comment by VWSCoronaDashboard18** [comment]
Nice that you did this one too!
- implemented usage of new utilities on the 'Ziekenhuizen en de zorg' page;

**Review by VWSCoronaDashboard21 — DISMISSED** [review]
Really like the persist state when toggling between the graphs on this page! :clap:

File: packages/app/src/utils/trim-leading-null-values.ts (hunk 7 above)
Comment (VWSCoronaDashboard26) [reply packages/app/src/utils/trim-leading-null-values.ts:13]: I personally think this approach allows for more granular 'plug-n-play' usage of either utility without having to extend the existing functionality or having to worry about figuring out how the functionality works when unaware of the 'directional' argument (either leading or trailing) you'd have to pass into the method. Basically, all functions are kept to a minimum/simpler version this way.

Comment (VWSCoronaDashboard18) [reply packages/app/src/utils/trim-leading-null-values.ts:13]: Good argument! Thanks for elaborating. I thought it was weird to have both functions as the original one is not being used anywhere. Therefore I thought it would make sense to just extend that functionality or replace it.

File: packages/app/src/utils/count-leading-null-values.ts [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+ * Return the number of leading null values from the given value array.
+ * If a property name is specified it is used to check each value item's
+ * property for null.
+ *
+ * @param values - The given array
+ * @param property - Optional property to check
+ * @returns The index of the last element in the array meeting the predicate
+ */
+export const countLeadingNullValues = <T>(values: T[], property?: keyof T) => {
+  const predicate = isPresent(property) ? (value: T) => isFilled(value[property]) : (value: T) => isFilled(value);
```
Comment (VWSCoronaDashboard18) [inline packages/app/src/utils/count-leading-null-values.ts:13]: You can maybe shorten this further by having one function and do the isPresent check inside that single function...

Because now you have two `(value: T) => {}` functions

Comment (VWSCoronaDashboard18) [reply packages/app/src/utils/count-leading-null-values.ts:13]: That would look something like this:

`const predicate = (value: T) => (isPresent(property) ? isFilled(value[property]) : isFilled(value));`

Comment (VWSCoronaDashboard18) [reply packages/app/src/utils/count-leading-null-values.ts:13]: Feel free to resolve it either way, depending on your perspective.

File: packages/app/src/utils/__tests__/trim-leading-null-values.spec.ts [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+TrimLeadingNullValues('should trim null values from an array of objects', () => {
+  const data = [{ test: null }, { test: null }, { test: 1 }, { test: 2 }];
+  assert.is(trimLeadingNullValues(data, 'test').length, 2);
+});
+
+TrimLeadingNullValues('should trim null values from an array of values', () => {
+  const data = [null, null, 1, 2];
+  assert.is(trimLeadingNullValues(data).length, 2);
+});
```
Comment (VWSCoronaDashboard18) [inline packages/app/src/utils/__tests__/trim-leading-null-values.spec.ts:15]: Would it be nice to extend this test by checking if it only removes the leading values? And not deletes other null values?

Maybe also include a test with no null values at all?

Feel free to resolve this comment if it would take to much time or is out of scope.

File: packages/app/src/utils/__tests__/count-leading-null-values.spec.ts [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+CountLeadingNullValues('should count the amount of leading null values in an array of values', () => {
+  const data = [{ test: null }, { test: null }, { test: 1 }, { test: 2 }];
+  assert.is(countLeadingNullValues(data, 'test'), 2);
+});
+
+CountLeadingNullValues('should count the amount of leading null values in an array of objects', () => {
+  const data = [null, null, 1, 2];
+  assert.is(countLeadingNullValues(data), 2);
+});
```
[the remark this answers is not in the document]
Comment (VWSCoronaDashboard27) [reply packages/app/src/utils/__tests__/count-leading-null-values.spec.ts:15]: I think it is a good suggestion.

**Comment by VWSCoronaDashboard27** [comment]
Looks good for me.

**Review by VWSCoronaDashboard27 — DISMISSED** [review]

File: packages/app/src/utils/count-leading-null-values.ts (hunk 8 above)
Comment (VWSCoronaDashboard26) [reply packages/app/src/utils/count-leading-null-values.ts]: Great idea! However, I've taken this one step further. Please see if you like the solution and resolve accordingly.

File: packages/app/src/utils/__tests__/count-leading-null-values.spec.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (VWSCoronaDashboard26) [reply packages/app/src/utils/__tests__/count-leading-null-values.spec.ts]: I've updated the test cases for this file and the other utilities. I personally do think this covers some nice and sufficient scenarios, so feel free to resolve this conversation if you agree.

File: packages/app/src/utils/count-leading-null-values.ts (hunk 8 above)
Comment (VWSCoronaDashboard18) [reply packages/app/src/utils/count-leading-null-values.ts]: Nice! Great imrpovement!

File: packages/app/src/utils/__tests__/count-leading-null-values.spec.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (VWSCoronaDashboard18) [reply packages/app/src/utils/__tests__/count-leading-null-values.spec.ts]: Nice ones!

File: packages/app/src/utils/__tests__/trim-leading-null-values.spec.ts (hunk 9 above)
Comment (VWSCoronaDashboard18) [reply packages/app/src/utils/__tests__/trim-leading-null-values.spec.ts]: Nice ones!

**Review by VWSCoronaDashboard18 — APPROVED** [review]
Looks wonderfull!

**Review by VWSCoronaDashboard28 — APPROVED** [review]
